### PR TITLE
Feature/implement delete in http artifact repository

### LINKED ISSUE #5996: [FR] Delete artifact over HTTP is not implemented
### Willingness to contribute

Yes. I can contribute this feature independently.

### Proposal Summary

The `http_artifact_repository` doesn't implement the `delete_artifacts` method, therefore it's not possible to permanently delete artifacts (for instance with the garbace collector `mlflow gc`).

### Motivation

> #### What is the use case for this feature?

Delete artifacts with e.g. `mlflow gc`.

> #### Why is this use case valuable to support for MLflow users in general?

The abstract `ArtifactRepository` defines a `delete_artifacts` method, which is not implemented in the `HttpArtifactRepository`. Making this endpoint available completes the API.

> #### Why is this use case valuable to support for your project(s) or organization?

We use the `mlflow gc` (experimental) command to clean deleted experiment data. Since we use the `--serve-artifacts` option on the server, the experiments use the `HttpArtifactRepository` to store the models. But because that doesn't have the `delete_artifacts` method implemented, the garbace collector doesn't work.

> #### Why is it currently difficult to achieve this use case?

We had to implement our own `gc` method.

### Details

_No response_

### What component(s) does this bug affect?

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by idlefella**
## Related Issues/PRs

Closes #5996

## What changes are proposed in this pull request?

Implements the `delete_artifacts` method in the `HttpArtifactRepository`.

## How is this patch tested?

I implemented a unittest for the method.

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Possibility to delete artifacts over http. The HttpArtifactRepository now implements the delete method, which allows users to delete artifacts even when they are served over http.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [x] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by idlefella** [comment]
Hi, thanks for running the tests. I will fix the issues by tomorrow evening.

**Comment by harupy** [comment]
> The CONTRIBUTING.rst still mentions protobuf 3.6.0.

Confirmed. `CONTRIBUTING.rst` needs to be updated.

**Comment by harupy** [comment]
The windows failure is unrelated.

File: mlflow/store/artifact/http_artifact_repo.py [hunk 1]
```diff
@@ -66,3 +66,11 @@ def _download_file(self, remote_file_path, local_path):
             chunk_size = 1024 * 1024  # 1 MB
             for chunk in resp.iter_content(chunk_size=chunk_size):
                 f.write(chunk)
+
+    def delete_artifacts(self, artifact_path=""):
```
Comment (harupy) [inline mlflow/store/artifact/http_artifact_repo.py:70]:
```suggestion
    def delete_artifacts(self, artifact_path=None):
```

**Comment by harupy** [comment]
@idlefella Can I push one commit to update [examples/mlflow_artifacts](https://github.com/mlflow/mlflow/tree/master/examples/mlflow_artifacts)?

File: mlflow/store/artifact/http_artifact_repo.py [hunk 2]
```diff
@@ -66,3 +66,11 @@ def _download_file(self, remote_file_path, local_path):
...
+        endpoint = posixpath.join("/", artifact_path)
+        resp = http_request(self._host_creds, endpoint, "DELETE", stream=True, timeout=10)
+        augmented_raise_for_status(resp)
+        if resp.status_code == 200:
+            return True
+        return False
```
Comment (harupy) [inline mlflow/store/artifact/http_artifact_repo.py:76]: Do we need to return True/False?

**Comment by idlefella** [comment]
> @idlefella Can I push one commit to update [examples/mlflow_artifacts](https://github.com/mlflow/mlflow/tree/master/examples/mlflow_artifacts)?

Sure, I added you as a collaborator in my fork. I think you should be able to push any commits.

File: mlflow/store/artifact/http_artifact_repo.py (hunk 2 above)
Comment (idlefella) [reply mlflow/store/artifact/http_artifact_repo.py:76]: No, I will remove it.

**Comment by harupy** [comment]
@idlefella Thanks, pushed a commit.

**Review by harupy — APPROVED** [review]
LGTM :) Thanks for the contribution!

**Pull request merged**